sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1122 @@
+            transformation = self.transformation_
+
+        elif isinstance(init, np.ndarray):
+            pass
+
+        elif init == 'pca':
+            pca = PCA(n_components=self.n_features_out,
+                      random_state=self.random_state_)
+            t_pca = time.time()
+            if self.verbose:
+                print('[{}] Finding principal components...'.format(
+                    self.__class__.__name__))
+                sys.stdout.flush()
+
+            pca.fit(X)
+            if self.verbose:
+                t_pca = time.time() - t_pca
+                print('[{}] Found principal components in {:5.2f}s.'.format(
+                    self.__class__.__name__, t_pca))
+
+            transformation = pca.components_
+
+        elif init == 'identity':
+            if self.n_features_out is None:
+                transformation = np.eye(X.shape[1])
+            else:
+                transformation = np.eye(self.n_features_out, X.shape[1])
+
+        return transformation
+
+    def _select_target_neighbors(self, X, y, n_targets, **kwargs):
+        """Find the target neighbors of each data sample.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels indices.
+
+        n_targets : int
+            The number of target neighbors to select for each sample in X.
+
+        Returns
+        -------
+        target_neighbors: array, shape (n_samples, n_targets)
+            An array of neighbors indices for each sample.
+        """
+
+        t_targets = time.time()
+        if self.verbose:
+            print('[{}] Finding the target neighbors...'.format(
+                self.__class__.__name__))
+            sys.stdout.flush()
+
+        target_neighbors = _select_target_neighbors(X, y, n_targets, **kwargs)
+
+        if self.verbose:
+            t_targets = time.time() - t_targets
+            print('[{}] Found the target neighbors in {:5.2f}s.'.format(
+                self.__class__.__name__, t_targets))
+
+        return target_neighbors
+
+    def _compute_grad_static(self, X, targets):
+        """Compute the gradient contributed by the target neighbors.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        targets : array, shape (n_samples, n_targets)
+            The k nearest neighbors of each sample from the same class.
+
+        Returns
+        -------
+        grad_targets, shape (n_features, n_features)
+            An array with the sum of all outer products of samples-targets.
+        """
+
+        t_grad_static = time.time()
+        if self.verbose:
+            print('[{}] Computing static part of the gradient...'.format(
+                self.__class__.__name__))
+
+        grad_targets = _compute_grad_static(X, targets)
```

Comment:
should we inline the function so that the reader doesn't need to jump around the file / not know whether it is used elsewhere?

Reply:
Done.